CONTRIBUTING.md

```diff
@@ -15,39 +15,141 @@ Otherwise, go read Docker's
 
 ### Development Environment Setup
 
-Swarm is written in [the Go programming language](http://golang.org) and manages its dependencies using [Godep](http://github.com/tools/godep). Go and Godep also use `git` to fetch dependencies.
-So, you are required to install the Go compiler, and the `git` client (e.g. `apt-get install golang git` on Ubuntu).
+Swarm is written in [the Go programming language](http://golang.org) and manages its dependencies using [Godep](http://github.com/tools/godep).  This guide will walk you through installing Go, forking the Swarm repo, building Swarm from source and contributing pull requests (PRs) back to the Swarm project.
 
-You may need to setup your path and set the `$GOPATH` and `$GOBIN` variables, e.g:
+#### Install git, Go and Godep
+If you don't already have `git` installed, you should install it.  For example, on Ubuntu:
 ```sh
-mkdir -p ~/projects/swarm
-cd ~/projects/swarm
-export GOPATH=$PWD
-export GOBIN=$GOPATH/bin
+sudo apt-get install git
 ```
 
-As previously mentioned, you need `godep` to manage dependencies for Swarm. The `godep` command can be installed using the following command:
+You also need Go 1.4+.  Download Go from [https://golang.org/dl/](https://golang.org/dl/).  To install on Linux:
+```sh
+tar xzvf go1.5.3.linux-amd64.tar.gz
+sudo mv go /usr/local
+```
+
+Create a go project directory in your home directory:
+```sh
+mkdir ~/gocode    # any name is fine
+```
+
+Add these to your `.bashrc`:
+```sh
+export GOROOT=/usr/local/go
+export GOPATH=~/gocode
+export PATH=$PATH:$GOPATH/bin:$GOROOT/bin
+```
+
+Close and reopen your terminal.
+
+Install Godep:
+
 ```sh
 go get github.com/tools/godep
 ```
-and you will find the `godep` binary under `$GOBIN`.
 
-Then you can fetch `swarm` code to your `$GOPATH`.
+#### Fork the Swarm repo
+
+You'll want to create a personal fork of the Swarm repo.  This allows you to make changes in your own area and then later open pull requests to propose your work to the main Swarm project.
+
+Go to the [docker/swarm repo](https://github.com/docker/swarm) in a browser and click the "Fork" button in the top right corner.  You'll end up with your form of Swarm in `https://github.com/<YOUR GITHUB USERNAME>/swarm`.
+
+Now clone your personal fork:
+
+```sh
+mkdir -p $GOCODE/src/github.com/<YOUR GITHUB USERNAME>/
+cd $GOCODE/src/github.com/<YOUR GITHUB USERNAME>/
+git clone https://github.com/<YOUR GITHUB USERNAME>/swarm.git
+cd swarm
+```
+
+#### Pull upstream changes into your fork
+
+Your personal fork of Swarm is a point-in-time snapshot of the main docker/swarm repo.  As docker/swarm moves forward, it will accumulate commits that are not in your fork.  To keep your fork up to date, you need to set docker/swarm as your upstream and then periodically rebase.
+
+Do this (one time only) to set your repo's upstream:
+
+```sh
+git remote add upstream https://github.com/docker/swarm.git
```

Comment:
@abronan I want to make sure I am understanding correctly what you are saying.

I did this:
```sh
mkdir -p $GOPATH/src/github.com/docker
cd $GOPATH/src/github.com/docker
git clone https://github.com/docker/swarm.git
git remote add mgoelzer https://github.com/mgoelzer/swarm.git
cd swarm
```

Ok, so now I have a clone of `docker/swarm` with two remotes:  `origin` and `mgoelzer`

Is the idea that now every push I do will be to `mgoelzer` instead of `origin`?  Like `git push mgoelzer master`?

For example:
```sh
git checkout -b issue-9999
[do some local edits and commits]
git push mgoelzer issue-9999
```

Do I have the right idea here?